jlm/llvm/opt/alias-analyses/Andersen.cpp

```diff
@@ -0,0 +1,514 @@
+    for (PointerObjectIndex idx = 0; idx < NodeCount(); idx++) {
+      if (memoryNodes[idx] == nullptr)
+        continue; // Skip all nodes that are not MemoryNodes
+
+      applyPointsToSet(memoryNodes[idx], idx);
+
+      if (GetNode(idx).HasEscaped())
+        escapedMemoryNodes.push_back(memoryNodes[idx]);
+    }
+
+    // Finally make all nodes marked as pointing to external, point to all escaped memory nodes in the graph
+    for (auto source : pointsToExternal) {
+      for (auto target : escapedMemoryNodes) {
+        source->AddEdge(*target);
+      }
+      // Add an edge to the special PtG node called "external" as well
+      source->AddEdge(pointsToGraph->GetExternalMemoryNode());
+    }
+
+    return pointsToGraph;
+  }
+};
+
+/**
+ * A constraint of the form:
+ * for all x in P(pointer1), make P(x) a superset of P(pointer2)
+ * Example of application is a store, e.g. when *pointer1 = pointer2
+ */
+class AllPointeesPointToSupersetConstraint final {
+  PointerObjectIndex Pointer1_;
+  PointerObjectIndex Pointer2_;
+
+public:
+  AllPointeesPointToSupersetConstraint(PointerObjectIndex pointer1, PointerObjectIndex pointer2)
+          : Pointer1_(pointer1), Pointer2_(pointer2) {}
+
+  bool
+  Apply(PointerObjectSet& set)
+  {
+    bool modified = false;
+
+    for (PointerObjectIndex x : set.GetPointsToSet(Pointer1_)) {
+      modified |= set.MakePointsToSetSuperset(x, Pointer2_);
+    }
+
+    // TODO: external and escaped flag stuff
+
+    return modified;
+  }
+};
+
+/**
+ * A constraint of the form:
+ * P(loaded) = union of P(x) for all x in P(pointer)
+ * Example of application is a load, e.g. when loaded = *pointer
+ */
+class SupersetOfAllPointeesConstraint final {
+  PointerObjectIndex Loaded_;
+  PointerObjectIndex Pointer_;
+
+public:
+  SupersetOfAllPointeesConstraint(PointerObjectIndex loaded, PointerObjectIndex pointer)
+          : Loaded_(loaded), Pointer_(pointer) {}
+
+  bool
+  Apply(PointerObjectSet& set)
+  {
+    bool modified = false;
+
+    for (PointerObjectIndex x : set.GetPointsToSet(Pointer_)) {
+      modified |= set.MakePointsToSetSuperset(Loaded_, x);
+    }
+
+    // TODO: external and escaped flag stuff
+
+    return modified;
+  }
+};
+
+using ConstraintVariant = std::variant<AllPointeesPointToSupersetConstraint, SupersetOfAllPointeesConstraint>;
+
+/**
+ * A class for adding and applying constraints to the points-to-sets of the PointerObjectSet
+ * Some constraints are applied immediately, while others are kept in lists for later constraint solving.
+ * Use Solve() to calculate the final points-to-sets.
+ *
+ * Constraints on the special external and escaped nodes are built in.
+ */
+class PointerObjectConstraintSet final
+{
+public:
+  explicit PointerObjectConstraintSet(PointerObjectSet& set) : Set_(set) {}
+
+  PointerObjectConstraintSet(const PointerObjectConstraintSet& other) = delete;
+
+  PointerObjectConstraintSet&
+  operator =(const PointerObjectConstraintSet& other) = delete;
```

Comment:
What about deleting move constructor and move assignment?

Reply:
I did not know that. Learned something. Thanks! Personally, I prefer to have it explicitly instead of remembering such rules.